### PR TITLE
feat: relaxing overlay mount restrictions

### LINKED ISSUE #1547: 放宽overlay挂载限制
问题描述
目前sealer build image时会检测是否支持overlay文件系统。当前默认通过modprobe overlay去检测，我发现在容器环境和wsl2都会导致最终保存失败(能够保存，但是拉取的镜像丢失，算作保存失败)。

解决方案：
去除overlay2.go中modprobe检测，仅使用/proc/filesystems去检测。并希望能够在不支持overlay的情况下，通过日志在控制台上显示。
https://github.com/sealerio/sealer/blob/cf9bdb1667d6d519aa0af63037999d73646dcbb4/utils/mount/overlay2.go#L49

### PR DESCRIPTION
**Pull request by YTGhost**
### Describe what this PR does / why we need it
relaxing overlay mount restrictions

### Does this pull request fix one issue?

Fixes #1547

### CODE REVIEW

**Comment by justadogistaken** [comment]
I'm considering if `modprobe overlay` failed, why do we allow the process to keep on. Why no just throw the error? I'm afraid that relaxing the restriction will incur more problems. And I think `modprobe overlay` is enough for now. If there are users need to relax the restriction. We can provide a env or option for user.
@YTGhost @ningpengtao-coder

File: utils/mount/overlay2.go [hunk 1]
```diff
@@ -46,15 +45,15 @@ func NewMountDriver() Interface {
...
-		return false
-	}
+func supportsOverlay() (flag bool) {
 	f, err := os.Open("/proc/filesystems")
 	if err != nil {
 		return false
 	}
 	defer func() {
+		if !flag {
+			logrus.Infof("The current environment does not support overlay")
```
Comment (justadogistaken) [inline utils/mount/overlay2.go:55]: Maybe warn would be better

Comment (YTGhost) [reply utils/mount/overlay2.go]: Yep, Thx!

Comment (YTGhost) [reply utils/mount/overlay2.go]: > I'm considering if `modprobe overlay` failed, why do we allow the process to keep on. Why no just throw the error? I'm afraid that relaxing the restriction will incur more problems. And I think `modprobe overlay` is enough for now. If there are users need to relax the restriction. We can provide a env or option for user. @YTGhost @ningpengtao-coder

I think using `/proc/filesystems` to detect supported filesystems seems to be sufficient in a linux environment? In wsl and container environments, `modprobe overlay` may not be supported for use.

**Comment by YTGhost** [comment]
> I'm considering if `modprobe overlay` failed, why do we allow the process to keep on. Why no just throw the error? I'm afraid that relaxing the restriction will incur more problems. And I think `modprobe overlay` is enough for now. If there are users need to relax the restriction. We can provide a env or option for user. @YTGhost @ningpengtao-coder

I think using `/proc/filesystems` to detect supported filesystems seems to be sufficient in a linux environment? In wsl and container environments, `modprobe overlay` may not be supported for use.

**Comment by justadogistaken** [comment]
@YTGhost Ok, I understand that. Thanks for this work.

File: utils/mount/overlay2.go [hunk 2]
```diff
@@ -46,15 +45,15 @@ func NewMountDriver() Interface {
...
-		return false
-	}
+func supportsOverlay() (flag bool) {
 	f, err := os.Open("/proc/filesystems")
 	if err != nil {
 		return false
 	}
 	defer func() {
+		if !flag {
+			logrus.Warnf("The current environment does not support overlay")
```
Comment (justadogistaken) [inline utils/mount/overlay2.go:55]: Change it to Warn insteand of Warnf.

Comment (YTGhost) [reply utils/mount/overlay2.go]: Done

File: utils/mount/overlay2.go [hunk 3]
```diff
@@ -46,15 +45,15 @@ func NewMountDriver() Interface {
 	return &Default{}
 }
 
-func supportsOverlay() bool {
```
Comment (justadogistaken) [inline utils/mount/overlay2.go:49]: Hang on a second.
I think there are two condition to detect the overlay support.
"modprobe overlay" || "/proc/filesystems".
I'm just afraid that some of cases that there will be false in "/proc/filesystems" case.

**Comment by justadogistaken** [comment]
Sorry, I misunderstand this issue.
So it could happen that "mobprobe overlay" returns true, but nothing in the "/proc/filesystems", am I right?

**Comment by justadogistaken** [comment]
There is a function "checkOverlaySupported" in sealer:main.
You can check this out. Maybe we can merge the current function and the "checkOverlaySupported". After this work committed.

**Comment by justadogistaken** [comment]
@YTGhost 
```
	s := bufio.NewScanner(f)
	for s.Scan() {
		if s.Text() == "nodev\toverlay" {
			return true
		}
	}
	return false
```
For this part. I guess we could read them all in.
And check if "overlay" exist?
Should "nodev\t" be included?

**Review by justadogistaken — APPROVED** [review]

**Comment by justadogistaken** [comment]
Please change `if s.Text() == "nodev\toverlay"` to strings.Containes, then I will merge the pr.
And we don't have to scan(). Just read the file content in.

**Comment by YTGhost** [comment]
> Sorry, I misunderstand this issue. So it could happen that "mobprobe overlay" returns true, but nothing in the "/proc/filesystems", am I right?

It could happen that "mobprobe overlay" returns false, but exist in the "/proc/filesystems"

**Comment by YTGhost** [comment]
> There is a function "checkOverlaySupported" in sealer:main. You can check this out. Maybe we can merge the current function and the "checkOverlaySupported". After this work committed.

Yep! we can try to merge them

**Comment by justadogistaken** [comment]
@YTGhost Okay. We can complete current work first. And file a another pr for merging two functions.

File: utils/mount/overlay2.go [hunk 4]
```diff
@@ -46,24 +44,18 @@ func NewMountDriver() Interface {
...
+		if !flag {
+			logrus.Warn("The current environment does not support overlay")
 		}
 	}()
-	s := bufio.NewScanner(f)
-	for s.Scan() {
-		if s.Text() == "nodev\toverlay" {
-			return true
-		}
+	if strings.Contains(string(content), "nodev\toverlay") {
```
Comment (justadogistaken) [inline utils/mount/overlay2.go:57]: “nodev\toverlay” => "overlay"

Comment (YTGhost) [reply utils/mount/overlay2.go]: Done

File: utils/mount/overlay2.go [hunk 5]
```diff
@@ -46,24 +44,18 @@ func NewMountDriver() Interface {
...
+		if !flag {
+			logrus.Warn("The current environment does not support overlay")
 		}
 	}()
-	s := bufio.NewScanner(f)
-	for s.Scan() {
-		if s.Text() == "nodev\toverlay" {
-			return true
-		}
+	if strings.Contains(string(content), "overlay") {
```
Comment (justadogistaken) [inline utils/mount/overlay2.go:57]: return strings.Contains(string(content), "overlay")

Comment (YTGhost) [reply utils/mount/overlay2.go]: Done

**Comment by justadogistaken** [comment]
LGTM! @YTGhost Thanks.

**Comment by justadogistaken** [comment]
I just created a ecs.
I do as follows:
1. cat /proc/filesystems, no overlay.
2. modprobe overlay.
3. cat /proc/filesystems, overlay appears.

So I recommend that add a line to the first line of this function.
But we don't have to catch the return.
```
_ = exec.Command("modprobe", "overlay").Run()
```

**Comment by YTGhost** [comment]
> I just created a ecs. I do as follows:
> 
> 1. cat /proc/filesystems, no overlay.
> 2. modprobe overlay.
> 3. cat /proc/filesystems, overlay appears.
> 
> So I recommend that add a line to the first line of this function. But we don't have to catch the return.
> 
> ```
> _ = exec.Command("modprobe", "overlay").Run()
> ```

Yep, I think we should add it to make sure the module is loaded？
Should I create a new branch to modify it?

**Comment by justadogistaken** [comment]
@YTGhost 
Yes.
File a new pr to fix this, plz.

**Pull request merged**